examples/file-sharing.rs

```diff
@@ -0,0 +1,639 @@
+        ))
+    }
+
+    #[derive(Clone)]
+    pub struct Client {
+        sender: mpsc::Sender<Command>,
+    }
+
+    impl Client {
+        /// Dial the given peer at the given address.
+        pub async fn dial(&mut self, peer_id: PeerId, peer_addr: Multiaddr) {
+            let (sender, receiver) = oneshot::channel();
+            self.sender
+                .send(Command::Dial {
+                    peer_id,
+                    peer_addr,
+                    sender,
+                })
+                .await
+                .unwrap();
+            receiver.await.unwrap();
+        }
+
+        /// Advertise the local node as the provider of the given file on the DHT.
+        pub async fn start_providing(&mut self, file_name: String) {
+            let (sender, receiver) = oneshot::channel();
+            self.sender
+                .send(Command::StartProviding { file_name, sender })
+                .await
+                .unwrap();
+            receiver.await.unwrap();
+        }
+
+        /// Find the providers for the given file on the DHT.
+        pub async fn get_providers(&mut self, file_name: String) -> HashSet<PeerId> {
+            let (sender, receiver) = oneshot::channel();
+            self.sender
+                .send(Command::GetProviders { file_name, sender })
+                .await
+                .unwrap();
+            receiver.await.unwrap()
+        }
+
+        /// Request the content of the given file from the given peer.
+        pub async fn request_file(&mut self, peer: PeerId, file_name: String) -> Option<String> {
+            let (sender, receiver) = oneshot::channel();
+            self.sender
+                .send(Command::RequestFile {
+                    file_name,
+                    peer,
+                    sender,
+                })
+                .await
+                .unwrap();
+            receiver.await.ok()
+        }
+
+        /// Respond with the provided file content to the given request.
+        pub async fn respond_file(&mut self, file: String, channel: ResponseChannel<FileResponse>) {
+            self.sender
+                .send(Command::RespondFile { file, channel })
+                .await
+                .unwrap();
+        }
+    }
+
+    async fn event_loop(
+        mut swarm: Swarm<ComposedBehaviour>,
+        mut command_receiver: mpsc::Receiver<Command>,
+        mut event_sender: mpsc::Sender<Event>,
+    ) {
+        let mut outstanding_dial: HashMap<_, oneshot::Sender<()>> = HashMap::new();
+        let mut outstanding_start_providing = HashMap::new();
+        let mut outstanding_get_providers: HashMap<_, oneshot::Sender<HashSet<PeerId>>> =
+            HashMap::new();
+        let mut outstanding_request_file: HashMap<_, oneshot::Sender<String>> = HashMap::new();
+
+        loop {
+            futures::select! {
+                event = swarm.next() => match event.expect("Swarm stream to be infinite.") {
```

Comment:
I don't find `select_next_some` intuitive and usually find myself going back to the method's documentation to make sure I understand all its implications. While a bit more verbose, users should already be familiar with `next` and `expect`. I don't feel strongly about it. Please do insist in case you do.

Reply:
I would agree with going for `futures::select!` and `swarm.next()`. This syntax is clearer and easier to understand, for someone who is not familiar with `futures`.

Reply:
In this particular case I consider it just to be syntactic sugar, which I mostly like very much given libp2p's verbosity :-)